### ISSUE TITLE
DispVM should be non-networked by default, or have a clear option to users to do so 

### ISSUE THREAD

**Issue by mfc** [QubesOS/qubes-issues#1121]
DispVMs are meant to be used to open untrusted files. However if that file can exfiltrate information (such as the fact it was opened), that can be seen as a security failure in some use cases.

DispVMs should be non-networked by default, and/or there should an option change this default if desired.

**Comment by unman** [comment]
Look at #862 for the rationale for the current position.
There is already an option to change the default in R3. Look in qubes manager on advanced tab for a VM, or use qvm-prefs.

**Comment by mfc** [comment]
ah I hadn't seen that option in Qubes manager Advanced tab, I was looking in global settings. 

I feel like the use-cases for inherited-network DispVMs are fewer (...printing?) than the use-cases for no-network DispVMs (opening untrusted files), so I would still submit to that being the default, with users being able to change it in the particular VM's settings as desired.

**Comment by marmarek** [comment]
One note - this shouldn't affect starting web browser in DispVM (from KDE/Xfce menu), of course.

**Comment by mfc** [comment]
hey all, updating this for R4.0. it continues to be the case it is much too difficult for a new user to have non-networked dispVMs by default, which is a common use-case for using Qubes ("don't be afraid to open those enticing attachments!").

from https://github.com/QubesOS/qubes-issues/issues/1988#issuecomment-445493373: 
> In Qubes 4.0 it's enough to set netvm of whatever default DispVM you have (fedora-29-dvm?) to none, so I consider it implemented.

the default DispVM is networked by default, is likely already used for other purposes (like opening suspicious links / visiting suspicious websites). so changing the networking of the default DispVM will likely break existing use-cases the user has for this DispVM.

to make explicit, here is what a user needs to do in order to be able to open a potential malicious file in a non-networked disposableVM:
1. search in the documentation and find how to [create new disposableVM template](https://www.qubes-os.org/doc/disposablevm-customization/#creating-a-new-disposablevm-template) with no netvm (as all default templates are networked) 
```
[user@dom0 ~]$ qvm-create --template debian-10 --label red offline-debian10-disposablevm-template
[user@dom0 ~]$ qvm-prefs offline-debian10-disposablevm-template template_for_dispvms True
[user@dom0 ~]$ qvm-features offline-debian10-disposablevm-template appmenus-dispvm 1
[user@dom0 ~]$ qvm-prefs offline-debian10-disposablevm-template netvm
```
2. set change their dispVM in other qubes to this new offline-debian10-disposablevm-template, or make it new default:
```
[user@dom0 ~]$ qubes-prefs default_dispvm offline-debian10-disposablevm-template
```
3. open that suspicious file via "View in DisposableVM"

what are options to make this simpler?

1. ~~[networking pop up](https://github.com/QubesOS/qubes-issues/issues/4006)~~ -- we want to avoid more pop-ups
2. ship with an additional dispvm template that is clearly offline/non-networked. potentially make it default dispvm template?
3. additional context menu options "View in offline DisposableVM" and "Edit in offline DisposableVM"
4. ?

for (2), why potentially make it default? when a user wants to view a potentially malicious link via dispvm, they will go to the "Q Menu", select a browser from a dispvm template, then once it loads, paste the URL and visit. they will not be visiting the URL via some context menu like "View in DisposableVM" etc. (please correct me if there is some way to do "View in DisposableVM" for urls...)

so i see the primary use of **default dispvm** functionality is in the context of **opening files**. and the default for a security- and privacy-focused OS should be that networking is not essential for files and we should avoid an "[insecure default](https://github.com/QubesOS/qubes-issues/issues/1988#issuecomment-218724766)".

**Comment by andrewdavidwong** [comment]
I agree that the common use cases should be securely configured by default. Why not have online and offline DisposableVM Templates by default, and use the following defaults:

- Email hyperlinks open in an online DisposableVM
- Email attachments open in an offline DisposableVM 
- Local files (in Nautilus) open in an offline DisposableVM

This is how I've had my system configured for years, and I find it works well (except for attachment opening currently being broken). This is also covered in our documentation, but I agree that it should be the default, since we specifically tout the email use cases as a benefit of Qubes.

**Comment by SvenSemmler** [comment]
> Why not have online and offline DisposableVM Templates by default

Very good idea!

> This is how I've had my system configured for years

Me too, in addition I have set the respective policies (OpenURL and
OpenInVM) to 'ask' which allows me to always choose:

URLs:

- open in online dvm (aka clearnet)
- open in tor dvm (i.e. whonix)
- open in already existing dispVM ("disp1234")
- open in specialized qube (i.e. the one where I'm logged into GitHub)

VMs:

- open in offline dvm
- open in already existing dispVM ("disp1234")

> I find it works well (except for attachment opening currently being 
> broken).

There is the option to add common file types (PDF, Office formats etc)
to the "Files & Attachments" preference in Thunderbird and hook it to
qvm-open-in-dvm. That works just as well as thunderbird-qubes did for me
especially in combination with the 'ask' policy.

> This is also covered in our documentation, but I agree that it should
> be the default, since we specifically tout the email use cases as a
> benefit of Qubes.

Couldn't agree more.

**Comment by neowutran** [comment]
If the offline disposable vm is added by default it will be great to also have specific qubes-rpc policy settings for it. 
To deny any qubes-related way to access network, ex: ``qvm-run-vm --dispvm curl https://neowutran.ovh`` and others.

**Comment by ben-grande** [comment]
Some remarks. Creating an alternative disposable template is not cumbersome:

```
qvm-clone default-dvm offline-default-dvm
qvm-prefs offline-default-dvm netvm ""
```

This can be added to the default installation, no hassle here.

Making it the global default_dispvm is one more command, but not having network is not the best for all use cases. I mostly don't open files in disposables manually (only in automated workflows) because I rarely need it, what I do need is to open links in disposables. I have configured my global default to not have a netvm, but I would not recommend for it to be the default because users workflow will break, new users won't understand why they don't have internet etc.

Anyway, I understand that the current way of using alternative disposable template [requires setting up the qrexec policy](https://doc.qubes-os.org/en/latest/user/advanced-topics/disposable-customization.html#open-a-link-in-a-disposable-based-on-a-non-default-disposable-template-via-command-line-from-app-qube).

One alternative, is modifying the qrexec policy rules for `qubes.OpenInVM` and `qubes.OpenURL`. The `OpenURL` is restricted to HTTP, HTTPS and FTP, so no issue of `qubes-open -> xdg-open` opening a file... unless the user messed with mime.

Untested idea:
```
qubes.OpenURL  * @anyvm @dispvm                     allow
qubes.OpenURL  * @anyvm @dispvm:offline-default-dvm allow
qubes.OpenInVM * @anyvm @dispvm:offline-default-dvm allow
qubes.OpenInVM * @anyvm @dispvm                     allow target=@dispvm:offline-default-dvm
```

I cannot make everyone happy with this, but I think it should be the default in order to make the system usable by default.

> If the offline disposable vm is added by default it will be great to also have specific qubes-rpc policy settings for it. To deny any qubes-related way to access network, ex: `qvm-run-vm --dispvm curl https://neowutran.ovh` and others.

You can already do that. If the policy I proposed above is added to `90-default.policy`, you can add your policy at `30-user` to override with:
```
qubes.OpenURL  * @anyvm @dispvm:offline-default-dvm allow
qubes.OpenInVM * @anyvm @dispvm:offline-default-dvm allow
*              * @anyvm @dispvm deny
```

Or you set the global disposable template to the `offline-default-dvm` and check if existing qubes are using the default on the per qube property.
